server/listener.h

```diff
@@ -74,76 +143,151 @@ class listener : public std::enable_shared_from_this<listener>
+    on_accept(boost::beast::error_code ec, tcp::socket socket)
     {
-        if (ec)
+        if(ec)
         {
-            fail(ec, "accept");
+            httpFail(ec, "listener_accept");
         }
         else
         {
-            session::make_session(
+            // Create the detector session and run it
+            std::make_shared<Detector<PlainSession, SslSession>>(
                 std::move(socket),
+                ctx_,
                 backend_,
                 subscriptions_,
                 balancer_,
-                dosGuard_);
+                dosGuard_)->run();
         }
 
         // Accept another connection
         do_accept();
     }
 };
 
-#endif  // LISTENER_H
+namespace Server
+{
+    using WebsocketServer = Listener<WsUpgrader, SslWsUpgrader>;
+    using HttpServer = Listener<HttpSession, SslHttpSession>;
+
+    static std::shared_ptr<WebsocketServer>
+    make_WebSocketServer(
+        boost::json::object const& config,
+        boost::asio::io_context& ioc,
+        ssl::context& ctx,
+        std::shared_ptr<BackendInterface> backend,
+        std::shared_ptr<SubscriptionManager> subscriptions,
+        std::shared_ptr<ETLLoadBalancer> balancer,
+        DOSGuard& dosGuard)
+    {
+        if (!config.contains("websocket_public"))
+            return nullptr;
+
+        auto const& wsConfig = config.at("websocket_public").as_object();
+
+        auto const address = 
+            boost::asio::ip::make_address(wsConfig.at("ip").as_string().c_str());
+        auto const port = 
+            static_cast<unsigned short>(wsConfig.at("port").as_int64());
+
+        auto server = std::make_shared<WebsocketServer>(
+            ioc,
+            ctx,
+            boost::asio::ip::tcp::endpoint{address, port},
+            backend,
+            subscriptions,
+            balancer,
+            dosGuard);
+
+        server->run();
+        return server;
+    }
+
+    static std::shared_ptr<HttpServer>
+    make_HttpServer(
+        boost::json::object const& config,
+        boost::asio::io_context& ioc,
+        ssl::context& ctx,
+        std::shared_ptr<BackendInterface> backend,
+        std::shared_ptr<SubscriptionManager> subscriptions,
+        std::shared_ptr<ETLLoadBalancer> balancer,
+        DOSGuard& dosGuard)
+    {
+        if (!config.contains("http_public"))
+            return nullptr;
+            
+        auto const& httpConfig = config.at("http_public").as_object();
+
+        auto const address = 
+            boost::asio::ip::make_address(httpConfig.at("ip").as_string().c_str());
+        auto const port = 
+            static_cast<unsigned short>(httpConfig.at("port").as_int64());
+
+        auto server = std::make_shared<HttpServer>(
+            ioc,
+            ctx,
+            boost::asio::ip::tcp::endpoint{address, port},
+            backend,
+            subscriptions,
+            balancer,
+            dosGuard);
+
+        server->run();
+        return server;
+    }
```

Comment:
I don't think we need two separate ports for HTTP and websocket. I think a single port should be able to handle all of the request types. The beast advanced flex server does this I'm pretty sure.